Reduce the information allocator need in createBlockMeta

LuQQiu opened the pull request:
In each `createBlockMetaInternal`, Space allocator will get a updated `BlockMetadataManagerView` which will do a full copy of pinList. Copy the pinList is slow when the list is huge (especially in async through) and copy is down in a `synchronized lock` which reduce the concurrency.  And actually, the allocator in the `createBlockMetaInternal` do not need the pin list info. It only needs the storage tier and directory information. 

This PR is to reduce the information given to allocator and thus reduce the contention and improve performance.

LuQQiu commented on core/server/worker/src/main/java/alluxio/worker/block/TieredBlockStore.java:
```diff
@@ -586,15 +585,15 @@ private TempBlockMeta createBlockMetaInternal(long sessionId, long blockId,
-          mAllocator.allocateBlockWithView(sessionId, initialBlockSize, location, getUpdatedView());
-      if (dirView == null) {
+      StorageDir dir = mAllocator.allocateBlockWithTierInfo(sessionId, initialBlockSize, location,
+          mMetaManager.getTiers(), mMetaManager.getAliasToTiersMap());
```
Do we also want to create another `BlockMetadataManagerView` to provide read only access to the tiers and directories?

madanadit replied:
+1. i like that better than exposing the tiers and alias to tiers maps

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/4330/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/4331/
Test FAILed.

LuQQiu commented:
@madanadit I added the `StorageMetadataView`, `StorageTierView`, and `StorageDirView` to only contains the information needed for allocator.

I renamed the original `StorageTierView` and `StorageDirView` to `StorageTierEvictableView` and `StorageDirEvictableView` to distinguish from the new ones.

In Allocator, there are many similar codes. I think of modifying the `StorageTierEvictableView` to extend `StorageTierView`, `StorageDirEvictableView` to extend `StorageDirView` to reduce the duplicate code, but then there will be many cast in our evictor. The evictor logics are much more complicated then allocator. I want to limit the impact of my changes as small as possible. 

PTAL, thanks!

madanadit commented on core/server/worker/src/main/java/alluxio/worker/block/TieredBlockStore.java:
```diff
@@ -586,8 +587,8 @@ private TempBlockMeta createBlockMetaInternal(long sessionId, long blockId,
-      StorageDirView dirView =
-          mAllocator.allocateBlockWithView(sessionId, initialBlockSize, location, getUpdatedView());
+      StorageDirView dirView = mAllocator.allocateBlockWithView(sessionId,
+          initialBlockSize, location, new StorageMetadataView(mMetaManager));
```
nit; did you consider calling `StorageMetadatView` `BlockMetadataManagerView` instead and renaming the existing `BlockMetadataManagerView` to `BlockMetadataManagerEvictableView` for consistency?

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/4335/
Test PASSed.

madanadit commented on core/server/worker/src/main/java/alluxio/worker/block/allocator/Allocator.java:
```diff
@@ -58,9 +76,9 @@ public static Allocator create(BlockMetadataManagerView view) {
    */
-  StorageDirView allocateBlockWithView(long sessionId, long blockSize, BlockStoreLocation location,
-      BlockMetadataManagerView view);
+  StorageDirEvictableView allocateBlockWithEvictableView(long sessionId, long blockSize,
```
hmm, does allocator need evict-able view?

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/4343/
Test FAILed.

LuQQiu commented on core/server/worker/src/main/java/alluxio/worker/block/meta/StorageTierView.java:
```diff
@@ -25,29 +23,36 @@
+   * @param tier which the tierView is constructed from
+   * @param evictable whether this class contains evict-related methods
+   */
+  public StorageTierView(StorageTier tier, boolean evictable) {
```
@madanadit Are there better ways than using a boolean?
Here i use `evictable` is because the `StorageDirView` and `StorageDirEvictableView` have different parameters.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/4345/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/4346/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/4349/
Test FAILed.

madanadit commented:
Thanks a lot @LuQQiu! Looks good after the comment / variable name fixes. @ggezer PTAL

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/4350/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/4351/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/4352/
Test PASSed.

ggezer commented on core/server/worker/src/main/java/alluxio/worker/block/evictor/LRFUEvictor.java:
```diff
@@ -78,9 +79,9 @@ public LRFUEvictor(BlockMetadataManagerView view, Allocator allocator) {
+    for (StorageTierView tier : mMetadataView.getTierViews()) {
       for (StorageDirView dir : tier.getDirViews()) {
-        for (BlockMeta block : dir.getEvictableBlocks()) {
+        for (BlockMeta block : ((StorageDirEvictorView) dir).getEvictableBlocks()) {
```
`StorageDirView` is casted to `StorageDirEvictorView` at many places. Can we avoid that somehow? Maybe `StorageTierEvictorView` could implement a one that returns `StorageDirEvictorView` instead.

LuQQiu replied:
I tried to use generic view of `StorageTierView` and `StorageDirView` and implement the abstract methods in allocator and evictor classes. The detail code listed in https://github.com/LuQQiu/alluxio/pull/1. This way reduce many unneeded cast.

However, in the allocators, we directly operate on abstract class `BlockMetadataView`.
And intellij will complain:
```
for (StorageTierView tierView : mMetadataView.getTierViews()) {
        for (StorageDirView dirView : tierView.getDirViews()) { <-- complain about incompatible types
          if (dirView.getMediumType().equals(mediumType)
```
Incompatible types: required object, but found StorageDirView

LuQQiu commented:
Mistakenly push one more commit, force push back to the commit that has been reviewed and approved

calvinjia commented:
alluxio-bot, merge this please.

calvinjia commented:
alluxio-bot, cherry-pick this to branch-2.0 please.
